man/mlgsc_train.1.md
% MLGSC(1) MLgsc User Manuals
% Thomas Junier
% December 16, 2015

# NAME

mlgsc_train - train a model for MLGSC, the maximum-likelihood general sequence classifier

# SYNOPSIS

mlgsc_train -h

mlgsc_train [*options*] **DNA|Prot** *multiple-alignment* *phylogeny* 

OR

mlgsc_train [*options*] **DNA|Prot** *multiple-alignment*

# DESCRIPTION

mlgsc_train trains a model for use with mlgsc(1). The arguments are as follows:

**DMA|Prot**
:   specifies whether the model is trained on nucleotide (DNA or RNA) or protein
    sequences.

*multiple-alignment*
:   the name of a file containing a Fasta-formatted mutliple alignment of the
    classifying region. In the first form (above), the Fasta header contains an
    ID and the taxon name (the phylogeny being supplied in a separate file). In
    the second form, the Fasta header also contains the taxon's complete
    anscestry (see below).
    
    In the first form, the Fasta headers should have the following format:
    
    `><ID> <taxon-name> [rest]`
    
    that is, an ID, followed by the target taxon's name, followed by optional
    free-form information; the ID, name and rest (if any) should be separated by
    whitespace. For example, the following header

    `>ID_5756 Clostridium rpoB`

    states that the associated sequence belongs to taxon _Clostridium_. The ID
    is not used, but it *can* be used for cross-validation (see
    *mlgsc_xval*(1)), and it is simpler for *mlgsc_train* and *mlgsc_xval* to
    use the same format.

    In the second form, the Fasta headers must have the following format:

    `><ID>;tax=d:<domain name>,p:<phylum name>,c:<class name>,o:<order name>,f:<family name>,g:<genus name>,s:<scientific name>;`

    e.g:

    `>Aquifex_aeolicus;tax=d:Bacteria,p:Aquificae,c:Aquificae,o:Aquificales,f:Aquificaceae,g:Aquifex,s:Aquifex aeolicus;`

    This is the format used by Utax
    (http://www.drive5.com/usearch/manual/tax_annot.htmlhttp://www.drive5.com/usearch/manual/tax_annot.html)
    
*phylogeny*
:   the name of a file containing a phylogeny of the target taxa (see option
    **-T**). The phylogeny can be either a tree (in Newick format) or a taxonomy
    file. The default is a tree.
    
    A tree should have its leaves labeled by reference taxa (but see option
    **-I**). Inner nodes can be labeled as well (this is encouraged). 
    
    A taxonomy file lists the taxonomic position of each target taxon, one
    taxon per line, from the most general to the most specific clade. Clade
    names are separated by semicolons. For genus _Clostridium_ this might look
    like:

    Bacteria; Firmicutes; Clostridia; Clostridiales; Clostridiaceae; Clostridium

    The taxonmy file is parsed into a tree structure by creating a tree node for
    each clade not already in the tree. This means that there should be no
    "gaps" in the taxonomy: for example given the two following lines

    ```
    Clostridiales; Peptococcaceae; Dehalobacter
    Clostridiales; Peptococcus
    ```
    
    Peptococcus will end up as sister node to Peptococcaceae, which is wrong.
    Instead, the taxonomy should read:
    
    ```
    Clostridiales; Peptococcaceae; Dehalobacter
    Clostridiales; Peptococcaceae; Peptococcus
    ```
# OUTPUT

A binary file containing the model. By default, the file's name derives from the
multiple alignment file's name, but it can be specified with option **-o**.

The program may also output warnings (unless verbosity is set to 0, see option
**-v**) in the following situations:

* if a taxon is found in the alignment but not in the phylogeny
* if a taxon is found in the phylogeny but not in the alignment

in both cases, the taxon is ignored.


# OPTIONS

-o, \--output-file *filename*
:   Set the name of the output file (model) to *filename*. By default, the
    output file's name is derived from the alignment file's name by removing any
    extension and adding `.bcls` (for Binary CLaSsifier). 
-v, \--verbosity **0|1|2**
:   Set the verbosity level. Level 0 is quiet: only fatal errors are reported.
    Level 1 (default) adds warning messages, and level 2 (verbose) adds
    information about the run, such as the parameters.
-W, \--no-Henikoff-Weighting
:   Do not apply Henikoff weighting. By default, the input sequences are
    weighted using Henikoff and Henikoff's algorithm [1]. Weighting often, but
    not always, improves accuracy by a few percentage points. Use mlgsc_xval(1)
    to estimate the effect of weighting.
-I, \--id-tree
:   The phylogeny is labeled by sequence ID instead of taxa. The use case for
    this option is when the phylogenetic tree has been computed directly from
    the alignment passed as second argument. This has the following
    consequences: (i) the tree is labeled by sequence ID instead of target
    taxon name; (ii) some taxa may not be monophyletic according to this
    particular tree; (iii) the tree may yet better reflect the phylogeny of the
    sequences at hand than would a phylogeny of the same taxa but computed from
    a different gene. The program then behaves as follows:
    
    1. from the headers of the alignment file, it constructs a (sequence ID ->
       taxon name) map.
    1. it relabels the input tree, replacing sequence IDs with the corresponding
       taxon names according to the map.
    1. it condenses any clade that consists entirely of the same taxon into a
       single leaf, labeled by that taxon
    1. if any taxon is found more than once (after condensing, thus indicating
       paraphyly), it gives a unique name to each leaf by appending a different
       number to each instance of a taxon that has multiple occurrences.

    For examle, imagine we have the following tree, where the leaves are
    labeled by sequence ID:

    ```
             /---------------+ ID1
     /-------+                    
     |       \---------------+ ID6
     |                            
    =+               /-------+ ID2
     |       /-------+            
     |       |       \-------+ ID3
     \-------+                    
             |       /-------+ ID4
             \-------+            
                     \-------+ ID5
    
    ```
    Suppose further that we have the following  ID -> taxon map:
    ```
    ID1 -> Solanum
    ID2 -> Atropa
    ID3 -> Datura
    ID4 -> Datura
    ID5 -> Mandragora
    ID6 -> Solanum
    ```

# EXIT STATUS

0 iff the program exited successfully.

# BUGS

* The program could try to determine if the sequences are DNA or protein

# EXAMPLES

Data for the examples are in the `data` directory of the distribution.

## Example 1

```
$ mlgsc_train Prot Spo0A.msa firmicute_genera.nw
```

This trains a model on the protein sequences of Spo0A (master regulator of
sporulation in Firmicutes), using Spo0A.msa and firmicutegenera.nw as
alignment and phylogeny, respectively. The output is a file named Spo0A.bcls.

## Example 2

```
$ mlgsc_train -o Firmicute_Spo0A Prot Spo0A.msa firmicute_genera.nw
```

As in Example 1, except that the model is named Firmicute_Spo0A (via option
**-o**).

# SEE ALSO

`mlgsc` (1), `mlgsc_xval` (1) and `mlgsc_dump` (1).

MLgsc is available form  <https://github.com/tjunier/mlgsc>

A technical description of how mlgsc_train trains the models is found in [2].
Please cite this paper if you use MLgsc in published research. Thanks!

# REFERENCES

[1] Henikoff S, Henikoff JG (1994) Position-based sequence weights. J Mol Biol 243(4): 574–57 pmid:7966282 doi: 10.1016/0022-2836(94)90032-9

[2] Junier T, Hervé V, Wunderlin T, and Junier P (2015) MLgsc: a
Maximum-Likelihood, general sequence classifier. PLoS One <http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0129384>



src/mlgsc_train.hs
module Main where

import Options.Applicative
import System.FilePath.Posix
import qualified Data.List as L
import qualified Data.Set as S
import qualified Data.Map.Strict as M
import Data.Binary (encodeFile, encode)
import Data.ByteString.Lazy (ByteString, unpack)
import qualified Data.Text as ST
import qualified Data.Text.Lazy as LT
import qualified Data.Text.IO as STIO
import qualified Data.Text.Lazy.IO as LTIO
import System.Environment (getArgs, getProgName)
import Data.Digest.Pure.CRC32
import Data.Tree

import MlgscTypes
import NewickDumper
import FastA (FastA, fastATextToRecords, utax2plain)
import Classifier (buildClassifier, Classifier, 
        StoredClassifier(..), Metadata(..))
import Alignment
import IDTree
-- TODO should replace (most of) the above with:
import API (rawTree, fastaRecsAndTree, treeStringFromFasta,
    otuAlignmentMap, parsePhyloFormat)

data Params = Params {
                optSmallProb        :: Double
                , optScaleFactor    :: Double
                , optOutFName       :: String
                , optVerbosity      :: Int
                , optNoHenikoffWt   :: Bool
                , optIDtree         :: Bool
                , optPhyloFormat    :: PhyloFormat 
                , molType           :: Molecule
                , posParams         :: [String]
                }

parseSmallProb :: Parser Double
parseSmallProb = option auto
                    (long "small-prob"
                    <> short 'p'
                    <> metavar "SMALL_PROB"
                    <> value 0.0001
                    <> help "small probability for absent residues")

parseScaleFactor :: Parser Double
parseScaleFactor = option auto
                    (long "scale-factor"
                    <> short 's'
                    <> metavar "SCALE_FACTOR"
                    <> value 1000.0
                    <> help "scale factor for log(frequencies)")

parseOutFName :: Parser String
parseOutFName = option str
                    (long "output-file"
                    <> short 'o'
                    <> metavar "OUTFILE"
                    <> value ""
                    <> help "name of the output classifier (default: derived from alignment)")

parseVerbosityLevel :: Parser Int
parseVerbosityLevel = option auto
                    (long "verbosity"
                    <> short 'v'
                    <> metavar "VERBOSITY LEVEL"
                    <> value 1
                    <> help "0: quiet, 1: normal, 2: verbose")

parseWeighting :: Parser Bool
parseWeighting =  switch (
                    short 'W'
                    <> long "no-Henikoff-weighting"
                    <> help "don't perform Henikoff weighting of input aln")

parseOptions :: Parser Params
parseOptions = Params
                <$> parseSmallProb
                <*> parseScaleFactor
                <*> parseOutFName
                <*> parseVerbosityLevel
                <*> parseWeighting
                <*> switch (
                    short 'I'
                    <> long "id-tree"
                    <> help "input tree labeled by seq ID, not taxa")
                <*> option (str >>= parsePhyloFormat) (
                    short 'T'
                    <> long "tree-file-format"
                    <> help "T)axonomy or N)ewick (default)"
                    <> value Newick )
                <*> argument auto (metavar "<DNA|Prot>")
                <*> many (argument str (metavar "<alignment file> [tree file]"))

parseOptionsInfo :: ParserInfo Params
parseOptionsInfo = info (helper <*> parseOptions) 
                    ( fullDesc
                    <> progDesc "train model from alignment and tree"
                    <> Options.Applicative.header
                        "mlgsc_train - model trainer for the ML general sequence classifier")


main :: IO ()
main = do
    params <- execParser parseOptionsInfo
    let alnFName = head $ posParams params
    fastAInput <-  LTIO.readFile $ alnFName
    -- putStrLn $ treeStringFromFasta $ fastATextToRecords fastAInput
    treeString <- if (length $ posParams params) == 2
        then do
            c <- readFile $ last $ posParams params
            return $ Just c
        else do
            return Nothing
    -- we have to return these at the same time because in some cases (e.g., ID
    -- tree) they depend on one another.
    let (fastaRecs, tree) = getFastaRecsAndTree params fastAInput treeString
    let otuAlnMap = otuAlignmentMap (optNoHenikoffWt params) fastaRecs
    let outputFileName = outFName (optOutFName params) alnFName
    runInfo params tree outputFileName
    possibleWarnings params tree otuAlnMap
    let classifier = buildClassifier
                        (molType params)
                        (optSmallProb params) (optScaleFactor params)
                        otuAlnMap tree
    cmdln <- getCmdLine
    -- TODO: the classifier gets encoded twice... this is inefficient.
    let cksum = crc32 $ encode classifier
    let sc = StoredClassifier classifier (Metadata cmdln cksum)
    encodeFile outputFileName sc

getFastaRecsAndTree ::
    Params -> LT.Text -> Maybe String -> ([FastA], Tree OTUName)
getFastaRecsAndTree params fastAInput mbTreeString =
    case mbTreeString of
        -- no separate phylogeny file: take it from Fasta header}
        Nothing             -> (map utax2plain fastaRecords,
                                    rawTree Taxonomy $
                                    treeStringFromFasta fastaRecords)
        -- get the phylogeny from the phylogeny file's contents, which may be
        -- Taxonomy or Newick (and in that case, a taxon tree or an ID tree)
        (Just treeString)   -> fastaRecsAndTree
                                    (optIDtree params)
                                    fastaRecords
                                    (rawTree (optPhyloFormat params) treeString)
    where   fastaRecords = fastATextToRecords fastAInput

runInfo :: Params -> NewickTree -> String -> IO ()
runInfo params tree outFname
    | optVerbosity params < 2   = do return ()
    | otherwise                 = do
        putStrLn $ unlines [
            "mlgsc_train - building model ",
            ("input alignment:  " ++ (head $ posParams params)),
            ("input tree: " ++ (last $ posParams params)),
            ("output: " ++ outFname),
            ("molecule: " ++ (show $ molType params)),
            ("small prob: " ++ (show $ optSmallProb params)),
            ("scale factor: " ++ (show $ optScaleFactor params)),
            ("Henikoff weighting: " ++ (show $ not $ optNoHenikoffWt params)),
            ("Tree format: " ++ (show $ optPhyloFormat params)),
            ("ID tree: " ++ (show $ optIDtree params)),
            if (optIDtree params)
                then "Relabeled tree: " ++ (ST.unpack $ treeToNewick tree)
                else ""
            ]

getCmdLine :: IO String
getCmdLine = do
            prog <- getProgName
            args <- getArgs
            return $ L.intercalate " " $ prog : args

possibleWarnings :: Params -> OTUTree -> AlnMap -> IO ()
possibleWarnings params tree otuAlnMap =
    if (optVerbosity params) > 0
        then do
            if not $ null $ treeOTUsNotInALn tree otuAlnMap
                then do
                    putStrLn "The following tree taxa  are NOT found in the alignment:"
                    mapM_ STIO.putStrLn $ treeOTUsNotInALn tree otuAlnMap
                else return ()
            if not $ null $ alnOTUsNotInTree tree otuAlnMap
                then do
                    putStrLn "The following alignment taxa  are NOT found in the tree:"
                    mapM_ STIO.putStrLn $ alnOTUsNotInTree tree otuAlnMap
                else return ()
        else return ()

-- returns a list of tree OTUs not found in the alignment

treeOTUsNotInALn :: OTUTree -> AlnMap -> [ST.Text]
treeOTUsNotInALn otuTree otuAlnMap =
    filter (\otu -> M.notMember otu otuAlnMap) $ fringe otuTree

-- returns a list of alignment OTUs not found in the tree

alnOTUsNotInTree :: OTUTree -> AlnMap -> [ST.Text]
alnOTUsNotInTree otuTree otuAlnMap =
    filter (\otu -> S.notMember otu treeOTUSet) $ M.keys otuAlnMap
    where   treeOTUSet = S.fromList $ fringe otuTree
    
-- Computes the name of the output file, based on the value of the "optOutFName"
-- option and possibly of the alignment file name.

outFName :: String -> String -> String
outFName optOutFN alnFN 
      | optOutFN == ""  = replaceExtension alnFN "bcls"
      | otherwise       = optOutFN
